File: app/src/main/java/ru/neosvet/vestnewage/service/CheckService.java
package ru.neosvet.vestnewage.service;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import androidx.core.app.NotificationCompat;
import androidx.lifecycle.LifecycleService;
import androidx.work.Configuration;
import androidx.work.Constraints;
import androidx.work.Data;
import androidx.work.ExistingWorkPolicy;
import androidx.work.NetworkType;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkContinuation;
import androidx.work.WorkManager;

import ru.neosvet.utils.Const;
import ru.neosvet.vestnewage.App;
import ru.neosvet.vestnewage.R;
import ru.neosvet.vestnewage.helpers.NotificationHelper;
import ru.neosvet.vestnewage.workers.CheckWorker;
import ru.neosvet.vestnewage.workers.LoaderWorker;

/**
 * Created by NeoSvet on 09.11.2019.
 */

public class CheckService extends LifecycleService {
    public static final String TAG = "CheckService";

    @Override
    public void onCreate() {
        super.onCreate();
    }

    public static void postCommand(boolean start) {
        Intent intent = new Intent(App.context, CheckService.class);
        intent.putExtra(Const.START, start);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            App.context.startForegroundService(intent);
        else
            App.context.startService(intent);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent == null || !intent.getBooleanExtra(Const.START, false)) {
            stopForeground(true);
            return Service.START_NOT_STICKY;
        }

        Context context = getApplicationContext();
        NotificationHelper notifHelper = new NotificationHelper();
        NotificationCompat.Builder notifBuilder = notifHelper.getNotification(
                context.getString(R.string.site_name),
                context.getString(R.string.check_new),
                NotificationHelper.CHANNEL_MUTE)
                .setProgress(0, 0, true);
        startForeground(NotificationHelper.NOTIF_CHECK, notifBuilder.build());

        try {
            Configuration configuration = new Configuration.Builder()
                    .setMinimumLoggingLevel(android.util.Log.INFO)
                    .build();
            WorkManager.initialize(getApplicationContext(), configuration);
        } catch (Exception e) {
            //e.printStackTrace();
        }

        Data data = new Data.Builder()
                .putString(Const.TASK, TAG).build();
        WorkManager work = WorkManager.getInstance(context);
        Constraints constraints = new Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED)
                .setRequiresBatteryNotLow(true)
                .build();
        OneTimeWorkRequest task = new OneTimeWorkRequest
                .Builder(CheckWorker.class)
                .setConstraints(constraints)
                .build();
        WorkContinuation job = work.beginUniqueWork(TAG,
                ExistingWorkPolicy.REPLACE, task);
        task = new OneTimeWorkRequest
                .Builder(LoaderWorker.class)
                .setInputData(data)
                .setConstraints(constraints)
                .build();
        job = job.then(task);
        job.enqueue();
        return super.onStartCommand(intent, flags, startId);
    }
}

File: app/src/main/java/ru/neosvet/vestnewage/service/CheckService.kt
package ru.neosvet.vestnewage.service

import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.lifecycle.LifecycleService
import ru.neosvet.vestnewage.App
import ru.neosvet.vestnewage.R
import ru.neosvet.vestnewage.data.NeoList
import ru.neosvet.vestnewage.helper.SummaryHelper
import ru.neosvet.vestnewage.loader.UpdateLoader
import ru.neosvet.vestnewage.network.NeoClient
import ru.neosvet.vestnewage.network.Urls
import ru.neosvet.vestnewage.utils.Const
import ru.neosvet.vestnewage.utils.Files
import ru.neosvet.vestnewage.utils.NotificationUtils

class CheckService : LifecycleService() {
    companion object {
        @JvmStatic
        fun postCommand(start: Boolean) {
            val intent = Intent(App.context, CheckService::class.java)
            intent.putExtra(Const.START, start)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
                App.context.startForegroundService(intent)
            else App.context.startService(intent)
        }
    }

    private var isRun = false

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent == null || !intent.getBooleanExtra(Const.START, false)) {
            if (Build.VERSION.SDK_INT == Build.VERSION_CODES.M)
                stopForeground(true)
            else stopForeground(STOP_FOREGROUND_REMOVE)
            return START_NOT_STICKY
        }
        if (isRun) return START_NOT_STICKY
        isRun = true
        initNotif()
        Thread { startLoad() }.start()
        return super.onStartCommand(intent, flags, startId)
    }

    private fun initNotif() {
        val utils = NotificationUtils()
        val notif = utils.getNotification(
            getString(R.string.app_name),
            getString(R.string.check_new),
            NotificationUtils.CHANNEL_MUTE
        )
            .setProgress(0, 0, true)
        startForeground(NotificationUtils.NOTIF_CHECK, notif.build())
    }

    private fun startLoad() {
        val loader = UpdateLoader(NeoClient())
        try {
            Urls.restore()
            val list = loader.checkSummary(true)
            val pref = getSharedPreferences(SummaryHelper.TAG, Context.MODE_PRIVATE)
            if (pref.getBoolean(Const.MODE, true)) {
                if (loader.checkAddition())
                    list.add(Pair(getString(R.string.new_in_additionally), Files.RSS + "1"))
            }
            if (pref.getBoolean(Const.DOCTRINE, false)) {
                if (loader.checkDoctrine())
                    list.add(Pair(getString(R.string.new_in_doctrine), Files.RSS + "2"))
            }
            if (pref.getBoolean(Const.PLACE, false)) {
                if (loader.checkAcademy())
                    list.add(Pair(getString(R.string.new_in_academy), Files.RSS + "3"))
            }
            if (list.isNotEmpty) pushNotification(list)
        } catch (ignored: Exception) {
        }
        loader.cancel()
        isRun = false
        postCommand(false)
    }

    private fun pushNotification(list: NeoList<Pair<String, String>>) {
        val helper = SummaryHelper()
        helper.preparingNotification()
        val several = list.size > 1
        if (several || list.first().second != Files.RSS)
            helper.updateBook()
        list.reset(true)
        list.forEach {
            helper.createNotification(it.first, it.second)
            if (several) {
                helper.muteNotification()
                helper.showNotification()
            }
        }
        if (several) helper.groupNotification()
        else helper.singleNotification(list.current().first)
        helper.setPreferences()
        helper.showNotification()
        list.clear()
    }
}